src/modules/rust.rs

```diff
@@ -1,601 +1,919 @@
+    rust_env_info: &RustToolingEnvironmentInfo,
+) -> Option<String> {
+    type Outcome = RustupRunRustcVersionOutcome;
+
+    match rust_env_info.get_rustup_rustc_version(context) {
+        Outcome::RustcVersion(rustc_version) => {
+            format_rustc_version(rustc_version, config.version_format)
+        }
+        Outcome::RustupNotWorking | Outcome::ToolchainUnknown => {
+            // If `rustup` can't be executed, or there is no environmental toolchain, we can
+            // execute `rustc --version` without triggering a toolchain download
+            format_rustc_version(&execute_rustc_version(context)?, config.version_format)
+        }
+        Outcome::ToolchainNotInstalled(name) => Some(name.to_string()),
+        Outcome::Err => None,
+    }
+}
+
+fn get_module_numeric_version(
+    context: &Context,
+    _config: &RustConfig,
+    rust_env_info: &RustToolingEnvironmentInfo,
+) -> Option<String> {
+    type Outcome = RustupRunRustcVersionOutcome;
+
+    match rust_env_info.get_rustup_rustc_version(context) {
+        Outcome::RustcVersion(version) => {
+            let release = version.split_whitespace().nth(1).unwrap_or(version);
+            Some(format_semver(release))
+        }
+        Outcome::RustupNotWorking | Outcome::ToolchainUnknown => {
+            let (numver, _toolchain) = rust_env_info.get_rustc_verbose_version(context)?;
+            Some(numver.to_string())
+        }
+        Outcome::ToolchainNotInstalled(_) | RustupRunRustcVersionOutcome::Err => None,
+    }
+}
+
+fn get_toolchain_version(
+    context: &Context,
+    _config: &RustConfig,
+    rust_env_info: &RustToolingEnvironmentInfo,
+) -> Option<String> {
+    type Outcome = RustupRunRustcVersionOutcome;
+
+    match rust_env_info.get_rustup_rustc_version(context) {
+        Outcome::RustcVersion(_) | Outcome::ToolchainNotInstalled(_) => {
+            let toolchain_override = rust_env_info
+                .get_env_toolchain_override(context)
+                // This match arm should only trigger if the toolchain override
+                // is not None because of how get_rustup_rustc_version works
+                .expect("Toolchain override was None: programming error.");
+            Some(toolchain_override.to_string())
+        }
+        Outcome::RustupNotWorking | Outcome::ToolchainUnknown => {
+            let default_host_triple = rust_env_info
+                .get_rustup_settings(context)
+                .default_host_triple();
+            let (_numver, toolchain) = rust_env_info.get_rustc_verbose_version(context)?;
+            Some(format_toolchain(toolchain, default_host_triple))
+        }
+        Outcome::Err => None,
+    }
+}
+
+fn env_rustup_toolchain(context: &Context) -> Option<String> {
+    log::trace!("Searching for rustup toolchain in environment.");
+    let val = context.get_env("RUSTUP_TOOLCHAIN")?;
+    Some(val.trim().to_owned())
+}
+
+fn execute_rustup_default(context: &Context) -> Option<String> {

```

Comment:
Is this still needed, or does `default_toolchain` from the config file handle this? But this could also be handled at a later date.

Reply:
I think overall, I'd like to clean up a lot of these and figure out a sane naming/organization scheme. For instance, right now, the method `get_env_toolchain_override` sounds like it's retrieving an override from environmental variables, but really it's attempting to follow `rustup`'s procedure, which involves reading from the environment, checking toolchain and local override files, etc.

With the way things are currently organized and named, I find it difficult to think about what's necessary and what's not.